### ISSUE TITLE
Parameter Substitution Cast Problem

### ISSUE THREAD

**Issue by burnes** [crate/crate#5176]
**CrateDB version**: 
1.0.5

**JVM version**:
8

**OS version / environment description**:
Linux

**Problem description**:

The IF and CASE statements require all returned values to be of the same type. In pure SQL this works fine, but if I'm using parameter substitution I have to cast values according to one data type.

As I can not send explicit LONG values (or no automatic cast is happening) I get the following error:

    ProgrammingError: SQLActionException[SQLParseException: if type of default result argument integer does not match type of results argument long]

**Use case**:

As we are using dynamic requests it's very unhandy to cast manually to the desired data type. It would be nice if Crate could just return the "highest" type needed from the query.

By the way this works in PostgresSQL.

**Steps to reproduce**:

This works perfectly:

```
CREATE TABLE test (vi INTEGER, vl LONG);
INSERT INTO test (vi, vl) VALUES (1,1), (2,2), (3,3)
SELECT vi, vl, IF(vi=1, vl, 33) FROM test LIMIT 100;
```

Error with Parameter substitution (in Crate Python):

```
from crate.client import connect

CRATE_HOST = '...'

conn = connect(CRATE_HOST, error_trace=True)
cursor = conn.cursor()
cursor.execute(
    """
    SELECT vi, vl, if(vi=?, vl, ?) from test limit ?;
    """,
    (1, 33, 100)
)
data = cursor.fetchall()
print(data)
```

**Comment by mfussenegger** [comment]
We currently don't have type inference / type precedence for function arguments but it's definetly something we want to address in the future.

(This is somewhat related to https://github.com/crate/crate/issues/4514)

**Comment by urbanhusky** [comment]
Is there a roadmap for this? This blocks our user-definable, dynamic queries.

**Comment by joemoe** [comment]
hey @urbanhusky, there is no eta for this at the moment.

**Comment by urbanhusky** [comment]
@joemoe We (Senseforce) need this urgently

**Comment by mfussenegger** [comment]
Short update here: The example in the initial issue description works now in master. This will be in the next feature release (2.3)

**Comment by matriv** [comment]
Feature is released with [2.3.0](https://crate.io/docs/crate/reference/en/latest/appendices/release-notes/2.3.0.html)
